### PR TITLE
Partial payment test

### PR DESCRIPTION
**Pull request by forestb**
Added an additional test where LevelUp was the second payment and the discount amount available covered the balance in full.

### CODE REVIEW

File: LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs [hunk 1]
```diff
@@ -297,5 +297,74 @@ public void SplitTenderExample_LevelUp_Then_Cash_With_Adjusted_Exemption_And_Adj
...
+
+            proposedOrderValues.ShouldBeEquivalentTo(expectedProposedOrderValues,
+                $"Expected: {expectedProposedOrderValues}" + Environment.NewLine +
+                $"Actual: {proposedOrderValues}");
+
+            // available discount amount $10.00
+            int availableDiscountAmount = 1000;
+
+            // POS applies $10.00 pretax discount to check and updates subtotal and tax
+            currentTax = (originalSubtotal - availableDiscountAmount) / TAX_RATE;
```
Comment (mmercurio) [inline LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs]: Why use `TAX_RATE` here to calculate tax, but not above?  I realize the end result is the same, but the style difference is odd and may lead one to wonder if there's a problem.

**Comment by forestb** [comment]
@mmercurio, there are other scenarios we should add to test cases.  I'll write up a ticket to add those.

File: LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs [hunk 2]
```diff
@@ -106,16 +95,21 @@ public void SplitTenderExample_Cash_Then_LevelUp()
...
 
-            int itemsSubtotalAmount = 2000;
-            int taxAmountDue = itemsSubtotalAmount / TAX_RATE;
+            Check check = new Check(originalTotalOwed, originalTaxOwed);
 
             // Cashier tenders $10 to cash
             // Updated check:
             //   PAID cash:     $10
             //   TOTAL DUE:     $12
+            check.ApplyTender(1000);
```
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs]: The `Check` class supports applying tenders.

File: LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs [hunk 3]
```diff
@@ -151,26 +132,22 @@ public void SplitTenderExample_Cash_Then_LevelUp()
             int availableDiscountAmount = 100;
 
             // POS applies pretax discount to check and updates subtotal and tax
-            itemsSubtotalAmount -= availableDiscountAmount;
-            totalTaxAmount = itemsSubtotalAmount / TAX_RATE;
-            totalOutstandingAmount = itemsSubtotalAmount + totalTaxAmount - paidCashAmount;
-            spendAmount = totalOutstandingAmount + availableDiscountAmount;
+            check.ApplyDiscount(availableDiscountAmount);
```
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs]: The `Check` class supports applying discounts.

**Comment by forestb** [comment]
@mmercurio, I've added a `Check` class to streamline some of the calculations and help make writing additional tests a bit more intuitive.  I've also updated the pre-existing tests to use the new class.  We were essentially simulating a POS already, and this just streamlines the logic.  Let me know what you think.

File: LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs (hunk 1 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs]: This should be addressed by the new `Check` class, but it was a good catch.

**Comment by forestb** [comment]
Also added [Jira ticket POS-363](https://levelup.atlassian.net/browse/POS-363) for adding some testing to edge cases where discount is applied and exemption amounts.

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿namespace LevelUp.Pos.ProposedOrders.Tests.Helpers
+{
+    /// <summary>
+    /// This class is meant to streamline expected calculations from a typical point of sale for the purpose of 
+    /// testing.  The behavior of this class expects a starting total amount due (including tax), and the tax amount.
+    /// It handles applying pre-tax discounts and standard payments/tenders.
+    /// </summary>
+    internal class Check
+    {
+        #region Private data members
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: lose the region

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        private int TotalTenders { get; set; }
+
+        private int AdjustedSubtotal => StartingSubtotal - TotalDiscounts;
+        private decimal AdjustedSubtotalInDollars => AdjustedSubtotal / 100.0M;
+        private int AdjustedTax => decimal.ToInt32(AdjustedSubtotalInDollars * TaxRate * 100.0M);
+
+        #endregion
+
+        /// <summary>
+        /// The current amount, in cents, owed by a customer, in cents, at any given moment.
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: what's with the echo?

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        /// </summary>
+        public int TotalOutstandingAmount => AdjustedSubtotal + AdjustedTax - TotalTenders;
+
+        /// <summary>
+        /// The total amount of tax due, in cents, on a check, in cents, at any given moment.
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: same deal

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        /// </summary>
+        public int TotalTaxAmount => AdjustedTax;
+
+        /// <summary>
+        /// By passing in a total amount owed and the tax due on a check, we can simulate a simplistic point of sale
+        /// for testing purposes.
+        /// </summary>
+        /// <param name="total">The intial amount due on the check, in cents.</param>
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: including tax, correct?

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    /// testing.  The behavior of this class expects a starting total amount due (including tax), and the tax amount.
+    /// It handles applying pre-tax discounts and standard payments/tenders.
+    /// </summary>
+    internal class Check
+    {
+        #region Private data members
+
+        private int StartingTotal { get; }
+        private int StartingTax { get; }
+        private int StartingSubtotal => StartingTotal - StartingTax;
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: Subtotal --> UntaxedTotal, maybe?

File: LevelUp.Pos.ProposedOrders.Tests/LevelUp.Pos.ProposedOrders.Tests.csproj [hunk 9]
```diff
@@ -48,6 +48,7 @@
   <ItemGroup>
     <Compile Include="CalculatorTests.cs" />
     <Compile Include="Data\CalculatorTestData.cs" />
+    <Compile Include="Helpers\Check.cs" />
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/LevelUp.Pos.ProposedOrders.Tests.csproj]: Helpers/ --> Mocks/ ?  Is this technically a mock?

File: LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs [hunk 10]
```diff
@@ -287,8 +248,69 @@ public void SplitTenderExample_LevelUp_Then_Cash_With_Adjusted_Exemption_And_Adj
...
+            // available discount amount $10.00
+            int availableDiscountAmount = 1000;
+
+            // POS applies $10.00 pretax discount to check and updates subtotal and tax
+            check.ApplyDiscount(availableDiscountAmount);
+
+            // Create completed order: expected values
+            AdjustedCheckValues expectedCompletedOrderValues =
+                new AdjustedCheckValues(
+                    spendAmount: 1100,      // spendAmount will include appliedDiscount and remaining tax
```
Comment (devlaf) [inline LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs]: I spent a bunch of time trying to figure out why this number is what it is.  Maybe worth adding a comment.

File: LevelUp.Pos.ProposedOrders.Tests/LevelUp.Pos.ProposedOrders.Tests.csproj (hunk 9 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/LevelUp.Pos.ProposedOrders.Tests.csproj]: Good call.

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs (hunk 4 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: Lost it.

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs (hunk 5 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: Roger, Roger.

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs (hunk 7 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: Yes, added to description.

File: LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs (hunk 10 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs]: ✔️

File: LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs (hunk 8 above)
Comment (forestb) [inline LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs]: I think the common language in the restaurant world is subtotal.  How strongly do you feel about this?  Is there a reason "Subtotal" is not correct?

**Comment by devlaf** [comment]
Well I'm happy with the changes.  LGTM
